File: internal/gitops/message_test.go
package gitops

import (
	"strings"
	"testing"
)

func TestParseTemplate(t *testing.T) {
	tests := []struct {
		name    string
		text    string
		wantErr bool
		wantSrc string
	}{
		{name: "an empty template compiles the shipped default", text: "  ", wantSrc: DefaultTemplate},
		{name: "the documented field form", text: `pmngr: update {{.ItemID}} "{{.Title}}"`, wantSrc: `pmngr: update {{.ItemID}} "{{.Title}}"`},
		{name: "the short lowercase form of the story", text: "{{action}} {{id}}: {{title}}", wantSrc: "{{action}} {{id}}: {{title}}"},
		{name: "an unbalanced action is refused", text: "{{action", wantErr: true},
		{name: "an unknown function is refused", text: "{{nosuchplaceholder}}", wantErr: true},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			tpl, err := ParseTemplate(tc.text)
			if tc.wantErr {
				if err == nil {
					t.Fatalf("ParseTemplate(%q) = nil error, want a failure", tc.text)
				}
				if CodeOf(err) != CodeTemplateInvalid {
					t.Errorf("code = %q, want %q", CodeOf(err), CodeTemplateInvalid)
				}
				return
			}
			if err != nil {
				t.Fatalf("ParseTemplate(%q): %v", tc.text, err)
			}
			if tpl.Text() != tc.wantSrc {
				t.Errorf("Text() = %q, want %q", tpl.Text(), tc.wantSrc)
			}
		})
	}
}

func TestTemplateRenderSubject(t *testing.T) {
	fields := Fields{
		ItemID:     "ACME-US-0042",
		Title:      "Login with SSO",
		Type:       "story",
		Status:     "in_progress",
		PrevStatus: "todo",
		ProjectKey: "ACME",
		Board:      "team-alpha",
		Action:     ActionMove,
		User:       "jose",
		Date:       "2026-09-04",
		Tool:       "gintrack 0.4.1 (companion)",
	}

	tests := []struct {
		name   string
		text   string
		fields Fields
		want   string
	}{
		{
			name:   "the shipped default",
			text:   "",
			fields: fields,
			want:   `pmngr: update ACME-US-0042 "Login with SSO"`,
		},
		{
			name:   "the four placeholders the story asks for",
			text:   "{{action}} {{id}}: {{title}}",
			fields: fields,
			want:   "move ACME-US-0042: Login with SSO",
		},
		{
			name:   "the type placeholder",
			text:   "{{action}} {{type}} {{id}}",
			fields: fields,
			want:   "move story ACME-US-0042",
		},
		{
			name:   "the conventional-commits variant of docs/07",
			text:   "docs({{.ProjectKey}}): update {{.ItemID}} — {{.Title}}",
			fields: fields,
			want:   "docs(ACME): update ACME-US-0042 — Login with SSO",
		},
		{
			name:   "status, board, user and date",
			text:   "{{.Action}} {{.ItemID}} to {{.Status}} ({{.Board}}) {{.User}} {{.Date}}",
			fields: fields,
			want:   "move ACME-US-0042 to in_progress (team-alpha) jose 2026-09-04",
		},
		{
			name:   "an absent action defaults to update",
			text:   "{{action}} {{id}}",
			fields: Fields{ItemID: "ACME-T-0001"},
			want:   "update ACME-T-0001",
		},
		{
			name:   "a newline in the title is folded into the subject",
			text:   "{{action}} {{id}}: {{title}}",
			fields: Fields{ItemID: "ACME-T-0002", Title: "one\ntwo", Action: ActionCreate},
			want:   "create ACME-T-0002: one two",
		},
		{
			name:   "a batch of several items uses the built-in bulk subject",
			text:   "",
			fields: Fields{Action: ActionUpdate, Count: 12},
			want:   "pmngr: update 12 items",
		},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			tpl, err := ParseTemplate(tc.text)
			if err != nil {
				t.Fatalf("ParseTemplate: %v", err)
			}
			msg, err := tpl.Render(tc.fields)
			if err != nil {
				t.Fatalf("Render: %v", err)
			}
			if msg.Subject != tc.want {
				t.Errorf("subject = %q, want %q", msg.Subject, tc.want)
			}
			if strings.Contains(msg.Subject, "\n") {
				t.Error("a subject must never contain a newline")
			}
		})
	}
}

func TestTemplateRenderBodyTrailers(t *testing.T) {
	tests := []struct {
		name   string
		fields Fields
		want   []string
		absent []string
	}{
		{
			name: "a status change reports the transition",
			fields: Fields{
				ItemID: "ACME-US-0042", Title: "Login with SSO", Type: "story",
				PrevStatus: "todo", Status: "in_progress",
				Tool: "gintrack 0.4.1 (companion)", Action: ActionMove,
			},
			want: []string{
				"Item: ACME-US-0042",
				"Type: story",
				"Status: todo -> in_progress",
				"Tool: gintrack 0.4.1 (companion)",
			},
		},
		{
			name: "an unchanged status is reported once",
			fields: Fields{
				ItemID: "ACME-US-0042", Type: "story",
				PrevStatus: "todo", Status: "todo", Tool: "gintrack dev",
			},
			want:   []string{"Status: todo"},
			absent: []string{"->"},
		},
		{
			name: "an agent write is attributed",
			fields: Fields{
				ItemID: "ACME-T-0311", Type: "task", Tool: "gintrack dev", Agent: "claude-code",
			},
			want: []string{"Agent: claude-code"},
		},
		{
			name:   "a bulk write reports the item count",
			fields: Fields{Action: ActionUpdate, Count: 12, Tool: "gintrack dev"},
			want:   []string{"Items: 12", "Tool: gintrack dev"},
			absent: []string{"Item: "},
		},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			msg, err := MustParseTemplate("").Render(tc.fields)
			if err != nil {
				t.Fatalf("Render: %v", err)
			}
			for _, want := range tc.want {
				if !strings.Contains(msg.Body, want) {
					t.Errorf("body is missing %q:\n%s", want, msg.Body)
				}
			}
			for _, absent := range tc.absent {
				if strings.Contains(msg.Body, absent) {
					t.Errorf("body should not contain %q:\n%s", absent, msg.Body)
				}
			}
		})
	}
}

func TestTemplateRenderTruncatesTheSubject(t *testing.T) {
	long := strings.Repeat("very long title ", 12)
	msg, err := MustParseTemplate("").Render(Fields{ItemID: "ACME-US-0042", Title: long})
	if err != nil {
		t.Fatalf("Render: %v", err)
	}
	if got := len([]rune(msg.Subject)); got != SubjectLimit {
		t.Errorf("subject length = %d, want %d: %q", got, SubjectLimit, msg.Subject)
	}
	if !strings.Contains(msg.Body, strings.TrimSpace(long)) {
		t.Errorf("the full title must survive in the body:\n%s", msg.Body)
	}
}

func TestMessageString(t *testing.T) {
	tests := []struct {
		name string
		msg  Message
		want string
	}{
		{name: "subject only", msg: Message{Subject: "pmngr: update X"}, want: "pmngr: update X"},
		{
			name: "subject and body are separated by a blank line",
			msg:  Message{Subject: "s", Body: "Item: X"},
			want: "s\n\nItem: X",
		},
	}
	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			if got := tc.msg.String(); got != tc.want {
				t.Errorf("String() = %q, want %q", got, tc.want)
			}
		})
	}
}

File: web/src/git/message.test.ts
import { describe, expect, it } from 'vitest';

import {
  CommitTemplateError,
  DEFAULT_COMMIT_TEMPLATE,
  SUBJECT_LIMIT,
  renderCommitMessage,
  validateCommitTemplate,
} from '@/git/message';

/**
 * These cases mirror `internal/gitops/message_test.go` one for one: the two
 * runtimes render commit messages independently (ADR-006) and the only thing
 * that keeps them in step is the format documented in docs/06 §3.3.
 */
describe('renderCommitMessage', () => {
  const fields = {
    itemId: 'ACME-US-0042',
    title: 'Login with SSO',
    type: 'story',
    status: 'in_progress',
    prevStatus: 'todo',
    projectKey: 'ACME',
    board: 'team-alpha',
    action: 'move',
    user: 'jose',
    date: '2026-09-04',
    tool: 'gintrack 0.4.1 (companion)',
  } as const;

  it.each([
    ['the shipped default', '', `pmngr: update ACME-US-0042 "Login with SSO"`],
    ['the four placeholders of the story', '{{action}} {{id}}: {{title}}', 'move ACME-US-0042: Login with SSO'],
    ['the type placeholder', '{{action}} {{type}} {{id}}', 'move story ACME-US-0042'],
    [
      'the conventional-commits variant',
      'docs({{.ProjectKey}}): update {{.ItemID}} — {{.Title}}',
      'docs(ACME): update ACME-US-0042 — Login with SSO',
    ],
    ['the field form', '{{.Action}} {{.ItemID}} to {{.Status}}', 'move ACME-US-0042 to in_progress'],
  ])('renders %s', (_name, template, want) => {
    expect(renderCommitMessage(template, fields).subject).toBe(want);
  });

  it('folds a multi-line title into one subject line', () => {
    const { subject } = renderCommitMessage('{{action}} {{id}}: {{title}}', {
      itemId: 'ACME-T-0002',
      title: 'one\ntwo',
      action: 'create',
    });
    expect(subject).toBe('create ACME-T-0002: one two');
    expect(subject).not.toContain('\n');
  });

  it('uses the built-in subject for a batch of several items', () => {
    expect(renderCommitMessage('', { action: 'update', count: 12 }).subject).toBe(
      'pmngr: update 12 items',
    );
  });

  it('truncates a long subject and keeps the full title in the body', () => {
    const long = 'very long title '.repeat(12);
    const message = renderCommitMessage(DEFAULT_COMMIT_TEMPLATE, {
      itemId: 'ACME-US-0042',
      title: long,
    });
    expect(message.subject).toHaveLength(SUBJECT_LIMIT);
    expect(message.body).toContain(long);
  });

  it('writes the machine-readable trailers of docs/06 §3.3', () => {
    const { body } = renderCommitMessage('', fields);
    expect(body).toContain('Item: ACME-US-0042');
    expect(body).toContain('Type: story');
    expect(body).toContain('Status: todo -> in_progress');
    expect(body).toContain('Board: team-alpha');
    expect(body).toContain('Tool: gintrack 0.4.1 (companion)');
  });

  it('reports an unchanged status once, without an arrow', () => {
    const { body } = renderCommitMessage('', {
      itemId: 'ACME-US-0042',
      status: 'todo',
      prevStatus: 'todo',
    });
    expect(body).toContain('Status: todo');
    expect(body).not.toContain('->');
  });
});

describe('validateCommitTemplate', () => {
  it.each([
    ['an empty template falls back to the default', ''],
    ['the short form', '{{action}} {{id}}: {{title}}'],
    ['the field form', 'pmngr: update {{.ItemID}} "{{.Title}}"'],
  ])('accepts %s', (_name, template) => {
    expect(() => {
      validateCommitTemplate(template);
    }).not.toThrow();
  });

  it.each([
    ['an unbalanced brace', '{{action'],
    ['an unknown placeholder', '{{nosuchplaceholder}}'],
  ])('refuses %s', (_name, template) => {
    expect(() => {
      validateCommitTemplate(template);
    }).toThrow(CommitTemplateError);
  });

  it('names the placeholders a user may use', () => {
    expect(() => {
      validateCommitTemplate('{{nope}}');
    }).toThrow(/action/);
  });
});
